mobius-coroutines/src/main/java/com/spotify/mobius/coroutines/CoroutinesSubtypeEffectHandlerBuilder.kt

```diff
@@ -23,68 +30,152 @@ import kotlin.reflect.KClass
+                        currentJob.cancel()
+                        currentJob = launch { function.invoke(effect as G, eventsChannel) }
+                    }
+                }
+            }
+
+        })
         return this
     }
 
-    fun addEffectHandler(kClass: KClass<out F>, function: suspend (F) -> Flow<E>) {
-        val previousValue = effectHandlersMap.put(kClass, function)
+    fun addEffectHandler(kClass: KClass<out F>, effectHandler: EffectHandler<F, E>) {
+        val previousValue = effectHandlersMap.put(kClass, effectHandler)
         if (previousValue != null) error("Trying to add more than one handler for the effect ${kClass.simpleName}")
     }
 
-    fun build(coroutineContext: CoroutineContext) = build(CoroutineScope(coroutineContext))
+    fun build(coroutineContext: CoroutineContext = EmptyCoroutineContext) = Connectable { eventConsumer ->
+        val scope = CoroutineScope(coroutineContext)
+        val eventsChannel = Channel<E>()
+        val subEffectChannels = ConcurrentHashMap<KClass<out F>, Channel<F>>()
+
+        // Connects the eventConsumer
+        scope.launch {
+            for (event in eventsChannel) {
+                if (isActive) eventConsumer.accept(event)
+            }
+        }
 
-    private fun build(scope: CoroutineScope) = Connectable { eventConsumer ->
         object : Connection<F> {
             override fun accept(effect: F) {
                 scope.launch {
-                    if (scope.isActive) {
-                        val effectHandler = effectHandlersMap[effect::class] ?: error("No effectHandler for $effect")
-                        effectHandler.invoke(effect).collect {
-                            if (scope.isActive) eventConsumer.accept(it)
+                    // Creates an effectChannel if this is the first time the effect is processed
+                    val subEffectChannel = subEffectChannels.computeIfAbsent(effect::class) {
+                        val subEffectChannel = Channel<F>()
+                        val effectHandler =
+                            effectHandlersMap[effect::class] ?: error("No effectHandler for $effect")
+                        // Connects the effectHandler if this is the first time the effect is processed
+                        scope.launch {
+                            if (isActive) effectHandler.handleEffects(subEffectChannel, eventsChannel)
                         }
+                        subEffectChannel
                     }
+
+                    if (isActive) subEffectChannel.send(effect)
                 }
             }
 
             override fun dispose() {
-                scope.cancel(CancellationException("Effect Handler disposed"))
+                scope.cancel("Effect Handler disposed")
+                eventsChannel.close()
+                subEffectChannels.forEachValue(1) { it.close() }
             }
         }
     }
+
+    sealed interface CancellationPolicy {
```

Comment:
Wouldn't `ExecutionPolicy` be a more appropriate name given the available implementations?

Reply:
Great suggestion ! I changed the name